lib/plugin_manager.rb

```diff
@@ -0,0 +1,213 @@
+    system "RAILS_ENV='#{@environment}' rake db:migrate"
+  end
+
+  def _revert_migrations(plugin)
+    # todo should we migrate for all envs?
+    migration_path = OpenProject::Application.config.paths['db/migrate'].select { |path| path.include?(plugin.name) }
+    ActiveRecord::Migrator.migrate migration_path, 0
+  end
+
+  def _assets_precompile
+    system "RAILS_ENV='production' rake assets:precompile"
+  end
+
+  def _assets_webpack
+    system "RAILS_ENV='production' rake assets:webpack"
+  end
+
+  def _assets_clobber
+    system "RAILS_ENV='production' rake assets:clobber"
+  end
+
+  def _remove_plugin_from_gemfile_plugins_file(plugin)
+    _remove_plugin_from_gemfile_plugins(plugin)
+    _write_to_gemfile_plugins_file
+    _remove_gemfile_plugins_if_empty
+  end
+
+  def _remove_plugin_from_gemfile_plugins(plugin)
+    gemfile_plugins_new = ''
+    plugins = [plugin].concat _dependencies_only_required_by(plugin)
+    @gemfile_plugins.each_line do |line|
+      gemfile_plugins_new << line if _line_contains_no_plugin?(line, plugins)
+    end
+    @gemfile_plugins = gemfile_plugins_new
+  end
+
+  def _dependencies_only_required_by(plugin)
+    # todo this needs to be addressed
+    dependencies = plugin.dependencies
+    dependencies.select { |dependency| dependency._not_needed_by_any_other_than?(plugin) }
+  end
+
+  def _line_contains_no_plugin?(line, plugins)
+    result = true
+    plugins.each do |plugin|
+      result = false if plugin.included_in?(line)
+    end
+    result
+  end
+
+  def _remove_gemfile_plugins_if_empty
+    FileUtils.rm(GEMFILE_PLUGINS_PATH) if @gemfile_plugins == ''
+  end
+end
+
+class Plugin
+  def self._available?(name)
+    _available_plugins[name]
+  end
+
+  def self._available_plugins
+    @available_plugins || _load_available_plugins
+  end
+
+  def self._load_available_plugins
+    unless File.exists?(PLUGINS_YML_PATH)
+      puts 'Could not find plugin list, abort!'
+      exit
+    end
+    @available_plugins = YAML.load_file(PLUGINS_YML_PATH)
+  end
+
+  attr_reader :name
+
+  def initialize(name)
+    if name == 'pdf-inspector'
+      @name = 'pdf-inspector'
+    else
+      unless Plugin._available?(name)
+        puts 'Could not find plugin, abort!'
+        exit
+      end
+      @name = name
+      @url = Plugin._available_plugins[name][:url]
+      @branch = Plugin._available_plugins[name][:branch]
+    end
+  end
+
+  def _not_needed_by_any_other_than?(plugin)
+    all_other_plugin_names = Plugin._available_plugins.inject([]) do |result, (other_name, _)|
+      plugin.name == other_name ? result : result << other_name
+    end
+    all_other_plugins = all_other_plugin_names.inject([]) do
+      |result, name| result << Plugin.new(name)
+    end
+    all_dependencies_from_other_plugins = all_other_plugins.inject([]) do |result, other_plugin|
+      result.concat other_plugin.dependencies
+    end
+    all_dependencies_from_other_plugins.any?{|dependency| dependency.name == name}
```

Comment:
Space missing to the left of {.<br>Space between { and | missing.<br>Space missing inside }.